source/mir/math/stat.d

```diff
@@ -1626,6 +1627,789 @@ unittest
+    x.put([0.0, 1, 2, 3, 4].sliced);
+    assert(x.sum == 15.0);
+    x.put(5);
+    assert(x.sum == 21.0);
+}
+
+version(mir_test)
+@safe pure @nogc nothrow
+unittest
+{
+    import mir.ndslice.slice: sliced;
+
+    MapSummator!("a + 1", double, Summation.pairwise) x;
+    static immutable a = [0.0, 1, 2, 3, 4];
+    x.put(a.sliced);
+    assert(x.sum == 15.0);
+    x.put(5);
+    assert(x.sum == 21.0);
+}
+
+version(mir_test)
+@safe pure
+unittest
+{
+    import mir.ndslice.slice: sliced;
+    import mir.ndslice.fuse: fuse;
+
+    MapSummator!("a + 1", double, Summation.pairwise) x;
+    auto a = [
+        [0.0, 1, 2],
+        [3.0, 4, 5]
+    ].fuse;
+    auto b = [6.0, 7, 8].sliced;
+    x.put(a);
+    assert(x.sum == 21.0);
+    x.put(b);
+    assert(x.sum == 45.0);
+}
+
+/++
+Variance algorithms.
+
+See Also:
+    $(WEB en.wikipedia.org/wiki/Algorithms_for_calculating_variance, Algorithms for calculating variance).
++/
+enum VarianceAlgo
+{
+    /++
+    Performs Welford's online algorithm for updating variance. When
+    calculating variance initially (not an update), the two-pass algorithm is 
+    used, whereby the input is first centered and then the sum of squares is 
+    calculated from that.    
+    +/
+    online,
+    
+    /++
+    Calculates variance using E(x^^2) - E(x)^2 (alowing for adjustments for 
+    population/sample variance). This algorithm can be numerically unstable.
+    +/
+    naive
+}
+
+private
+@safe pure nothrow @nogc
+T square(T)(scope const T x)
+{
+    import mir.internal.utility: isFloatingPoint;
+    import mir.math.common: powi;
+
+    static if (isFloatingPoint!T) {
+        return powi(x, 2);
+    } else static if (__traits(compiles, {
+        T val = T.init * T.init;
+    })) {
+        return x * x;
+    } else {
+        static assert(0, "square: cannot square type: " ~ T.stringof);
+    }
+}
+
+@safe pure nothrow @nogc
+unittest {
+    int x = square(4);
+    assert(x == 16);
+    double y = square(5.0);
+	assert(y == 25);
+    
+    static assert(is(typeof(square(4)) == int));
+    static assert(is(typeof(square(5.0)) == double));
+}
+
+private
+mixin template variance_ops(T, Summation summation)
+{
+
+
+    ///
+    MeanAccumulator!(T, summation) meanAccumulator;
+
+    ///
+    size_t count() @property
+    {
+        return meanAccumulator.count;
+    }
+}
+
+///
+struct VarianceAccumulator(T, VarianceAlgo varianceAlgo, Summation summation)
+    if (isMutable!T && varianceAlgo == VarianceAlgo.naive)
+{
+    mixin variance_ops!(T, summation);
+
+    ///
+    MapSummator!(square, T, summation) sumOfSquares;
+    
+    ///
+    this(Range)(Range r)
+        if (isIterable!Range)
+    {
+        this.put(r);
+    }
+    
+    ///
+    this()(T x)
+    {
+        this.put(x);
+    }
+    
+    ///
+    void put(Range)(Range r)
+        if (isIterable!Range)
+    {
+        static if (hasShape!Range)
+        {
+            meanAccumulator.put(r);
+            sumOfSquares.put(r);
```

Comment:
2x `move()`

Reply:
@9il I hopefully have addressed this in the second commit.

Reply:
> 2x `move()`

@jmh530 this was a wrong comment, it should be only one `.move` for the last call. My bad.

In other hand. It's very preferred to evaluate the `r` range only once, because it may be a lazy range, for example, `someStringArray.as!double`